### PR TITLE
fix zero bin in categorical split

### LINKED ISSUE #3273: Results depend on categorical labels
Python package
LightGBM component: 
lightgbm==2.3.1
## Environment info
Operating System:
Windows 10
Python version: 3.7.6

The results (for example, cv) depend on categorical labels. As an example, I took Boston dataset, added artificial category feature with values ['low', 'mid', 'high']. Then, I encoded it:
dataset_1 is created based on {'low': 0, 'mid': 1, 'high': 2} encoding, dataset_2 is created based on {'low': 1, 'mid': 0, 'high': 2}.
And CV results are different.

```
import sklearn
from sklearn import datasets
import pandas

import lightgbm

lgb_params = {
    'boosting_type': 'gbrt',
    'learning_rate': 0.1,
    'max_depth': -1,
    'min_child_samples': 20,
    'min_child_weight': 0.001,
    'min_split_gain': 0.0,
    'n_jobs': -1,
    'num_leaves': 1023,
    'reg_alpha': 0.0,
    'reg_lambda': 0.0,
    'subsample': 1.0,
    'subsample_for_bin': 5,
    'subsample_freq': 0,
    'verbose': -1,
    'seed': 42
}

boston_dataset = sklearn.datasets.load_boston()

df_data = pandas.DataFrame(boston_dataset['data'], columns=boston_dataset['feature_names'])
df_target = pandas.Series(boston_dataset['target'])

df_data['category'] = pandas.cut(df_target, bins=[0, 20, 30, 80], labels=['low', 'mid', 'high'])
df_data_1 = df_data.copy()
df_data_1['category'] = df_data_1['category'].replace({'low': 0, 'mid': 1, 'high': 2}).astype(int)

for _ in range(10):
    dataset_1 = lightgbm.Dataset(data=df_data_1, label=boston_dataset['target'], categorical_feature=['category'])
    cv_result = lightgbm.cv(lgb_params, dataset_1, num_boost_round=500, nfold=2, seed=23, early_stopping_rounds=15,  
            metrics=['l1'], stratified=False)

    metric_name = next(iter(cv_result.keys()))
    print(len(cv_result[metric_name]), cv_result[metric_name][-1])

```

> 91 2.982193080452038
> 91 2.982193080452038
> 91 2.982193080452038
> 91 2.982193080452038
> 91 2.982193080452038
> 91 2.982193080452038
> 91 2.982193080452038
> 91 2.982193080452038
> 91 2.982193080452038
> 91 2.982193080452038

```
df_data_2 = df_data.copy()
df_data_2['category'] = df_data_2['category'].replace({'low': 1, 'mid': 0, 'high': 2}).astype('int')
for _ in range(10):
    dataset_2 = lightgbm.Dataset(data=df_data_2, label=boston_dataset['target'], categorical_feature=['category'])
    cv_result = lightgbm.cv(lgb_params, dataset_2, num_boost_round=500, nfold=2, seed=23, early_stopping_rounds=15,  
            metrics=['l1'], stratified=False)

    metric_name = next(iter(cv_result.keys()))
    print(len(cv_result[metric_name]), cv_result[metric_name][-1])
```

> 76 3.20606765690021
> 76 3.20606765690021
> 76 3.20606765690021
> 76 3.20606765690021
> 76 3.20606765690021
> 76 3.20606765690021
> 76 3.20606765690021
> 76 3.20606765690021
> 76 3.20606765690021
> 76 3.20606765690021

### PR DESCRIPTION
**Pull request by guolinke**
to fix #3273

### CODE REVIEW

File: src/io/dense_bin.hpp [hunk 1]
```diff
@@ -318,7 +318,9 @@ class DenseBin : public Bin {
     data_size_t gt_count = 0;
     data_size_t* default_indices = gt_indices;
     data_size_t* default_count = &gt_count;
-    if (Common::FindInBitset(threshold, num_threahold, most_freq_bin)) {
+    int8_t offset = most_freq_bin == 0 ? 1 : 0;
+    if (most_freq_bin > 0 &&
+        Common ::FindInBitset(threshold, num_threahold, most_freq_bin)) {
```
Comment (guolinke) [inline src/io/dense_bin.hpp:323]:
```suggestion
        Common::FindInBitset(threshold, num_threahold, most_freq_bin)) {
```

File: src/treelearner/feature_histogram.hpp [hunk 2]
```diff
@@ -1172,6 +1172,7 @@ class HistogramPool {
         }
       }
     }
+    Log::Warning("o0 %d, bin %d", offsets->at(0), num_total_bin);
```
Comment (guolinke) [inline src/treelearner/feature_histogram.hpp:1175]:
```suggestion
```

File: src/treelearner/feature_histogram.hpp [hunk 3]
```diff
@@ -649,16 +651,14 @@ class FeatureHistogram {
     double gain_shift = GetLeafGainGivenOutput<true>(
         sum_gradient, sum_hessian, meta_->config->lambda_l1, meta_->config->lambda_l2, parent_output);
     double min_gain_shift = gain_shift + meta_->config->min_gain_to_split;
-    bool is_full_categorical = meta_->missing_type == MissingType::None;
-    int used_bin = meta_->num_bin - 1 + is_full_categorical;
-    if (threshold >= static_cast<uint32_t>(used_bin)) {
+    if (threshold >= static_cast<uint32_t>(meta_->num_bin) || threshold < 0) {
```
Comment (guolinke) [inline src/treelearner/feature_histogram.hpp:654]:
```suggestion
    if (threshold >= static_cast<uint32_t>(meta_->num_bin)) {
```

Comment (guolinke) [inline src/treelearner/feature_histogram.hpp:654]:
```suggestion
    if (threshold >= static_cast<uint32_t>(meta_->num_bin) || threshold < meta_->offset) {
```

File: src/treelearner/feature_histogram.hpp [hunk 4]
```diff
@@ -649,16 +651,14 @@ class FeatureHistogram {
     double gain_shift = GetLeafGainGivenOutput<true>(
         sum_gradient, sum_hessian, meta_->config->lambda_l1, meta_->config->lambda_l2, parent_output);
     double min_gain_shift = gain_shift + meta_->config->min_gain_to_split;
-    bool is_full_categorical = meta_->missing_type == MissingType::None;
-    int used_bin = meta_->num_bin - 1 + is_full_categorical;
-    if (threshold >= static_cast<uint32_t>(used_bin)) {
+    if (threshold >= static_cast<uint32_t>(meta_->num_bin) || threshold < meta_->offset) {
```
Comment (guolinke) [inline src/treelearner/feature_histogram.hpp:654]:
```suggestion
    if (threshold >= static_cast<uint32_t>(meta_->num_bin) || threshold == 0) {
```

**Comment by guolinke** [comment]
@jameslamb 
any ideas about `/Users/runner/work/LightGBM/LightGBM/lightgbm.Rcheck/00_pkg_src/lightgbm/src/src/lightgbm_R.cpp:679:41: warning: cast between incompatible function types from 'LGBM_SER* (*)(LGBM_SE, LGBM_SE, LGBM_SE)' {aka 'LGBM_SER* (*)(LGBM_SER*, LGBM_SER*, LGBM_SER*)'} to 'DL_FUNC' {aka 'void* (*)()'} [-Wcast-function-type]` ?

**Comment by jameslamb** [comment]
I think it can safely be ignored, and would be fixed by https://github.com/microsoft/LightGBM/issues/3016.

The failing checks in this PR are because of this

```text
* checking installed package size ... NOTE
  installed size is  5.2Mb
  sub-directories of 1Mb or more:
    libs   4.5Mb
```

I think it is ok to ignore that NOTE, and that can be done by increasing the limit here: https://github.com/microsoft/LightGBM/blob/69a2691a599df7b259f2e79dc562895c8396c7d5/.ci/test_r_package.sh#L202-L206. Since we had this problem in another PR too (https://github.com/microsoft/LightGBM/pull/3264#issuecomment-671509757), I'll just make that change right now in a separate PR. Otherwise we will keep hitting this into one of these is merged.
